### PR TITLE
Support fields option on source-less indices

### PR DESCRIPTION
**Pull request by nik9000**
This adds support for fetching `fields` from doc values for a few field
types:
* long
* int
* short
* byte
* double
* short
* date
* boolean
* ip
* keyword
* unsigned long
* version string
* wildcard
* scaled float

We've had support for fetching from doc values forever using the
`docvalue_fields` option and this just plugs into that when the source
is disabled. It's possible that it's faster to plug into it in other
cases but we don't know those cases yet.

It does that by creating a `ValueFetcherSource` which is a place where
we can attach additional fetch machinery. At this point it mostly just
has the machinery for the `fields` fetch API and highlighters. But we
believe we'll want this spot for TSDB's synthetic source project. And so
we can make the fields API more efficient when fetching small numbers of
fields later.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/index/mapper/ValueFetcherSource.java [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+
+import org.elasticsearch.core.Nullable;
+import org.elasticsearch.index.query.SearchExecutionContext;
+
+import java.util.List;
+
+/**
+ * Resolves a {@link ValueFetcher} for a particular field.
+ */
+public interface ValueFetcherSource {
```
Comment (javanna) [inline server/src/main/java/org/elasticsearch/index/mapper/ValueFetcherSource.java:19]: having source in this name, given that value fetchers may retrieve fields from _source, is super confusing :)

File: server/src/main/java/org/elasticsearch/index/mapper/ValueFetcherSource.java [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+    ValueFetcher preferStored();
```
Comment (javanna) [inline server/src/main/java/org/elasticsearch/index/mapper/ValueFetcherSource.java:20]: stored here makes me think of stored fields

File: server/src/main/java/org/elasticsearch/search/fetch/subphase/FetchDocValuesPhase.java [hunk 3]
```diff
@@ -46,9 +46,9 @@ public FetchSubPhaseProcessor getProcessor(FetchContext context) {
             }
             ValueFetcher fetcher = new DocValueFetcher(
                 ft.docValueFormat(fieldAndFormat.format, null),
-                context.searchLookup().getForField(ft)
+                context.getSearchExecutionContext().getForField(ft)
             );
-            fields.add(new DocValueField(fieldAndFormat.field, fetcher));
+            fields.add((new DocValueField(fieldAndFormat.field, fetcher)));
```
Comment (javanna) [inline server/src/main/java/org/elasticsearch/search/fetch/subphase/FetchDocValuesPhase.java:51]: how are these two changes required?

**Comment by javanna** [comment]
I see that the new abstraction allows consumers to pull fields from different data sources. I think it would help to have a high-level description on which component pulls what from.

**Comment by cbuescher** [comment]
Sorry for the drive-by commenting on this issue. I read it and it looks like a great addition to the current "fields" behaviour. I was wondering how this plays along with CCS and some of its bwc guarantees? The way I understand it the new behaviour returns certain fields from 8.2+ indices that we wouldn’t from earlier indices, is that so? That means if a user requests a source-less field from both a 8.2 and 8.1 cluster connected via CCS, some documents returned might not show that field. 
I think thats perfectly fine, I just want to raise it here as a consideration given we talked about CCS bwc a lot recently.

File: server/src/main/java/org/elasticsearch/search/fetch/subphase/FetchDocValuesPhase.java (hunk 3 above)
Comment (nik9000) [reply server/src/main/java/org/elasticsearch/search/fetch/subphase/FetchDocValuesPhase.java:51]: Must be leftovers. Sorry about htat.

File: server/src/main/java/org/elasticsearch/index/mapper/ValueFetcherSource.java (hunk 2 above)
Comment (nik9000) [reply server/src/main/java/org/elasticsearch/index/mapper/ValueFetcherSource.java:20]: Maybe a bad name. I figured that _source is a stored field. OTOH maybe it's useful to have three layers - `_source`, stored fields, and doc values. Not sure.

File: server/src/main/java/org/elasticsearch/index/mapper/ValueFetcherSource.java (hunk 1 above)
Comment (nik9000) [reply server/src/main/java/org/elasticsearch/index/mapper/ValueFetcherSource.java:19]: Fair! Got any better ideas?

**Comment by javanna** [comment]
> Could you say that again with other words? I sort of half understand what you are getting at and when I try and fill in the blank I worry I'll get your meaning wrong.

sorry @nik9000 for being cryptic. I meant that I could use a summary in the description of the PR about the three different supported scenarios (the three methods in the new abstraction) and what the consumers are for each one.

**Comment by jtibshirani** [comment]
Sorry for my own drive-by comment, I found the logic in the last PR (#85185) a lot more focused and easy to understand. Would it make sense to start with that approach, then later see if we need this flexibility and "machinery" as your TSDB research evolves?

**Comment by nik9000** [comment]
> Sorry for my own drive-by comment, I found the logic in the last PR (#85185) a lot more focused and easy to understand. Would it make sense to start with that approach, then later see if we need this flexibility and "machinery" as your TSDB research evolves?

I'd be fine with that. As of today I don't think tsdb is going to need this directly.

This all might be more useful once we have synthetic source. Then you could `preferDocValues` to avoid having to load _source at all.

**Comment by javanna** [comment]
@nik9000 shall we close this?

**Comment by nik9000** [comment]
Yeah. ESQL does this pretty effectively as this point. I wasn't going to get this up to date with `main`.

**Pull request closed without merging**